### PR TITLE
Exceptions ftp

### PR DESCRIPTION
**Pull request by jim-parry**
Add exception handling to the Ftp library.

### CODE REVIEW

File: system/libraries/Ftp.php [hunk 1]
```diff
@@ -656,12 +656,13 @@ public function close()
...
 	 * @param	string	$line
 	 * @return	void
+	 * @throws	RuntimeException	In case of failure
 	 */
 	protected function _error($line)
 	{
 		$CI =& get_instance();
 		$CI->lang->load('ftp');
-		show_error($CI->lang->line($line));
+		throw new RuntimeException('Ftp: ' . $CI->lang->line($line));
```
Comment (narfbg) [inline system/libraries/Ftp.php]: We don't put spaces around concatenation.

I'm also not sure about this whole thing here ... translated exceptions smell, we may remove the translations.

**Comment by jim-parry** [comment]
Translated exception/error messages ... if these are meant to be recoverable, I don't see the issue. How is this different from the other libraries (eg email & form_validation) that have similar provisions?

**Comment by narfbg** [comment]
Form validation errors are displayed to the end user, exceptions shouldn't be.

Don't know about Email, haven't looked into it really, but it wouldn't surprise me if somebody made error translations there for "consistency" with everything else - similar to what we have here, bad decisions have been made in the past.
Unless the application itself is primarily an FTP client, there's no reason to trigger fatal errors and/or to show these errors to the end user at all.

**Comment by jim-parry** [comment]
Restarting this, since the feature/exceptions branch is too stale.

**Pull request closed without merging**